src/main/scala/mesosphere/marathon/api/v2/LabelSelector.scala

```diff
@@ -0,0 +1,84 @@
+package mesosphere.marathon.api.v2
+
+import mesosphere.marathon.state.AppDefinition
+
+import scala.util.control.NonFatal
+import scala.util.parsing.combinator.RegexParsers
+
+case class LabelSelector(key: String, fn: String => Boolean, value: List[String]) {
+  def matches(app: AppDefinition): Boolean = app.labels.contains(key) && fn(app.labels(key))
+
+}
+case class LabelSelectors(selectors: Seq[LabelSelector]) {
+  def matches(app: AppDefinition): Boolean = selectors.forall(_.matches(app))
+}
+
+/**
+  * Parse a label selector query.
+  * A label selector query has this format:
+  * Query: Selector {, Selector}
+  * Selector: ExistenceSelector | SetSelector | EqualsSelector
+  * ExistenceSelector: Id
+  * SetSelector: Label in|notin Set
+  * EqualsSelector: Label ==|!= Ident
+  * Set: ( Ident {, Ident} )
+  * Label: any character without whitespace
+  * Ident: any character without , or )
+  *
+  * Examples:
+  * test == foo
+  * test != foo
+  * environment in (production, qa)
+  * tier notin (frontend, backend)
+  * partition
+  * !!! in (---, &&&, +++)
+  * a in (t1, t2, t3), b notin (t1), c, d
+  *
+  */
+class LabelSelectorParsers extends RegexParsers {
+
+  def label: Parser[String] = """[^\s!=]+""".r
+  def eqOp: Parser[String] = """(==|!=)""".r
+  def setOp: Parser[String] = """(in|notin)""".r
+  def ident: Parser[String] = """[^,)=]+""".r
```

Comment:
I would prefer another name and another definition ;)

One idea would be to allow two syntaxes, literalLabel and quotedLabel. literalLabel only allows stuff like permissible in Java identifiers. quotedLabel follows the syntax for Java Strings. What do you think about that?